tensorflow_datasets/structured/movielens.py

```diff
@@ -0,0 +1,441 @@
+the values and the corresponding ranges are:
+  - 1: "Under 18"
+  - 18: "18-24"
+  - 25: "25-34"
+  - 35: "35-44"
+  - 45: "45-49"
+  - 50: "50-55"
+  - 56: "56+"
+- "user_occupation_label": the occupation of the user who made the rating
+represented by an integer-encoded label; labels are preprocessed to be
+consistent across different versions
+- "user_occupation_text": the occupation of the user who made the rating in
+the original string; different versions can have different set of raw text
+labels
+- "user_zip_code": the zip code of the user who made the rating
+
+In addition, the "100k-ratings" dataset would also have a feature "raw_user_age"
+which is the exact ages of the users who made the rating
+
+Datasets with the "-movies" suffix contain only "movie_id", "movie_title", and
+"movie_genres" features.
+"""
+
+_FORMAT_VERSIONS = ['25m', 'latest-small', '20m', '100k', '1m']
+_TABLE_OPTIONS = ['movies', 'ratings']
+
+
+class MovieLensConfig(tfds.core.BuilderConfig):
+  """BuilderConfig for MovieLens dataset."""
+
+  def __init__(
+      self,
+      format_version: Optional[str] = None,
+      table_option: Optional[str] = None,
+      download_url: Optional[str] = None,
+      parsing_fn: Optional[Callable[
+          [str],
+          Iterator[Tuple[int, Dict[str, Any]]],
+      ]] = None,
+      **kwargs
+  ) -> None:
+    """Constructs a MovieLensConfig.
+
+    Args:
+      format_version: a string to identify the format of the dataset, one of
+          '_FORMAT_VERSIONS'.
+      table_option: a string to identify the table to expose, one of
+          '_TABLE_OPTIONS'.
+      download_url: a string url for downloading the dataset.
+      parsing_fn: a callable for parsing the data.
+      **kwargs: keyword arguments forwarded to super.
+
+    Raises:
+      ValueError: if format_version is not one of '_FORMAT_VERSIONS' or if
+          table_option is not one of '_TABLE_OPTIONS'.
+    """
+    if format_version not in _FORMAT_VERSIONS:
+      raise ValueError('format_version must be one of %s.' % _FORMAT_VERSIONS)
+    if table_option not in _TABLE_OPTIONS:
+      raise ValueError('table_option must be one of %s.' % _TABLE_OPTIONS)
+    super(MovieLensConfig, self).__init__(**kwargs)
+    self._format_version = format_version
+    self._table_option = table_option
+    self._download_url = download_url
+    self._parsing_fn = parsing_fn
+
+  @property
+  def format_version(self) -> str:
+    return self._format_version
+
+  @property
+  def table_option(self) -> str:
+    return self._table_option
+
+  @property
+  def download_url(self) -> str:
+    return self._download_url
+
+  @property
+  def parsing_fn(self) -> Optional[Callable[
+      [str],
+      Iterator[Tuple[int, Dict[str, Any]]],
+  ]]:
+    return self._parsing_fn
+
+
+class MovieLens(tfds.core.GeneratorBasedBuilder):
+  """MovieLens rating dataset."""
+
+  BUILDER_CONFIGS = [
+      MovieLensConfig(
```

Comment:
What is the point of having the movies only without the rating ? Why not merging `25m-ratings` and `25m-movies` in a single config ?

Reply:
The purpose of having `25m-movies` separately is to give user access to the movies table so they can choose candidates from the set of all movies for recommendation. After joining the movie data and the rating data, movies would be repeated multiple times since each movie can have multiple ratings. Therefore it would be complicated for users to extract a set of movies with no duplicates.

Reply:
We need the `movies` dataset in order to compute retrieval evaluation metrics. For each query-candidate pair (from the `ratings` dataset), we'd like to compute how that candidate is ranked for that query _relative to all other candidates_. To get all other candidates, we need a deduplicated set of movies. Asking the users to do this manually would be complex and error prone, and so we provide this as a separate dataset.

Reply:
Thanks for the clarification